src/beam_search_scorer_cuda.cpp

```diff
@@ -72,15 +73,25 @@ void BeamSearchScorer_Cuda::Process(Sequences_Cuda& sequences,
 
 bool BeamSearchScorer_Cuda::IsDoneLater() const {
   cudaEventSynchronize(event_process_complete_);
+  // auto temp_state_gpu_ = CudaMallocHostArray<cuda::BeamScorerState>(1);
+  // cudaMemcpyAsync(temp_state_gpu_.get(), state_gpu_.get(), sizeof(cuda::BeamScorerState), ::cudaMemcpyDeviceToHost, stream_);
   return state_cpu_->not_done_count_ == 0;
 }
 
 void BeamSearchScorer_Cuda::Finalize(Sequences_Cuda& sequences,
-                                     size_t num_return_sequences,
-                                     std::span<int32_t> output,           // Word IDs of each sequence, with shape (batch_size * num_return_sequences, max_sequence_length)
-                                     std::span<float> sequence_scores) {  // Score of each sequence, with shape (batch_size * num_return_sequences).
-  assert(!output.empty());
-  cuda::LaunchBeamSearchScorer_Finalize(state_cpu_->batch_size_, *state_gpu_, sequences.GetSequences(), sequences.GetSequenceLength(), beam_hyps_, next_beam_scores_, output, sequence_scores, stream_);
+                                     size_t num_return_sequences) {
+  cuda::LaunchBeamSearchScorer_Finalize(state_cpu_->batch_size_, *state_gpu_, sequences.GetSequences(), sequences.GetSequenceLength(), beam_hyps_, next_beam_scores_, stream_);
+}
+
+RoamingArray<int32_t> BeamSearchScorer_Cuda::GetBeamHypothesis(size_t batch_id, size_t beam_id) const {
+  cuda_host_unique_ptr<int32_t*> hypothesis_ptr = CudaMallocHostArray<int32_t*>(1);
+  cuda_host_unique_ptr<int> hypothesis_length = CudaMallocHostArray<int>(1);
+  cuda_host_unique_ptr<float> hypothesis_score = CudaMallocHostArray<float>(1);
+  cuda::LaunchBeamSearchScorer_GetHypothesisPtr(batch_id, beam_id, beam_hyps_, hypothesis_ptr.get(), hypothesis_length.get(), hypothesis_score.get(), stream_);
+  CudaCheck() == cudaStreamSynchronize(stream_);
+  // std::cout << "hypothesis length: " << *hypothesis_length.get() << std::endl;
```

Comment:
nit: remove this unnecessary std::cout. You can use the DumpTensor to dump the result, just like this:
https://github.com/microsoft/onnxruntime-genai/blob/00ceb80e1984c408459dfabe92a5b4eb97318578/src/models/model.cpp#L56-L60